syncopy/connectivity/wilson_sf.py

```diff
@@ -0,0 +1,221 @@
+    '''
+
+    nSamples = CSD.shape[1]
+    
+    # perform ifft to obtain gammas.
+    gamma = np.fft.ifft(CSD, axis=0)
+    gamma0 = gamma[0, ...]
+    
+    # Remove any assymetry due to rounding error.
+    # This also will zero out any imaginary values
+    # on the diagonal - real diagonals are required for cholesky.
+    gamma0 = np.real((gamma0 + gamma0.conj()) / 2)
+
+    # check for positive definiteness
+    eivals = np.linalg.eigvals(gamma0)
+    if np.all(np.imag(eivals) == 0):    
+        psi0 = np.linalg.cholesky(gamma0)
+    # otherwise initialize with 1's as a fallback
+    else:
+        psi0 = np.ones((nSamples, nSamples))
+        
+    return psi0.T
+    
+
+def _plusOperator(g):
+
+    '''
+    The []+ operator from definition 1.2,
+    given by explicit Fourier transformations
+
+    The nFreq x nChannel x nChannel matrix `g` is given 
+    in the frequency domain.
+    '''
+
+    # 'negative lags' from the ifft
+    nLag = g.shape[0] // 2
+    # the series expansion in beta_k 
+    beta = np.fft.ifft(g, axis=0)
+
+    # take half of the zero lag
+    beta[0, ...] = 0.5 * beta[0, ...]
+    g0 = beta[0, ...].copy()
+
+    # Zero out negative lags
+    beta[nLag + 1:, ...] = 0
+
+    gp = np.fft.fft(beta, axis=0)
+
+    return gp, g0
+
+
+# --- End of Wilson's Algorithm ---
+
+
+def regularize_csd(CSD, cond_max=1e6, eps_max=1e-3, nSteps=50):
+
+    '''
+    Brute force regularize CSD matrix
+    by inspecting the maximal condition number 
+    along the frequency axis.
+    Multiply with different epsilon * I, 
+    starting with epsilon = 1e-10 until the
+    condition number is smaller than `cond_max`.
+    Raises a ValueError if the maximal regularization 
+    factor `epx_max` was reached but `cond_max` still not met.
+
+
+    Parameters
+    ----------
+    CSD : 3D :class:`numpy.ndarray`
+        The cross spectral density matrix
+        with shape (nFreq, nChannel, nChannel)
+    cond_max : float
+        The maximal condition number after regularization
+    eps_max : float
+        The largest regularization factor to be used. If
+        also this value does not regularize the CSD up
+        to `cond_max` a ValueError is raised.
+    nSteps : int
+        Number of steps between 1e-10 and eps_max.
+
+    Returns
+    -------
+    CSDreg : 3D :class:`numpy.ndarray`
+        The regularized CSD matrix with a maximal
+        condition number of `cond_max`
+    eps : float
+        The regularization factor used
+
+    '''
+
+    epsilons = np.logspace(-10, np.log10(eps_max), 25)
+    I = np.eye(CSD.shape[1])
+
+    CondNum = np.linalg.cond(CSD).max()
+
+    # nothing to be done
+    if CondNum < cond_max:
+        return CSD, 0
+
+    for eps in epsilons:        
+        CSDreg = CSD + eps * I
+        CondNum = np.linalg.cond(CSDreg).max()
+
+        if CondNum < cond_max:
+            return CSDreg, eps
+
+    msg = f"CSD matrix not regularizable with a max epsilon of {eps_max}!"
+    raise ValueError(msg)
```

Comment:
I think for the pure-backend it's nice to be syncopy agnostic, also for people contributing code?!

Reply:
Hm, okay. In general, I think it would be preferable if all Syncopy-related errors a user sees are actually `SPY*`-exceptions (this way we can also be sure that any incurring exception was indeed raised by us).  However, as discussed above, the chance that the CSD matrix is too irregular for Wilson is very low anyway, so the probability that a Syncopy user ever sees this error should be close to zero.